design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+4. After refactor, the startup time for applications built with this library should not increase. This will be measured by the startup time of [this](https://github.com/maplibre/maplibre-gl-native/blob/main/platform/ios/platform/ios/iosapp%20UITests/iosapp_UITests.swift) iOS test app.
+5. When implemented with a given style and set of tilesets, the compiled library should not decrease the rendering frame rate.
+6. This refactor will not introduce data loading or other bottlenecks that will negatively impact rendering performance.
+7. Developers currently using this library will not see any breaking API changes, and will be able to seamlessly upgrade versions.
+
+## Proposed Change
+
+### Shader(Program) Representation
+
+_Addresses core functionality [#1](#core) and [#2](#core)._
+
+We need an external representation for programs (shaders) so they can be added or replaced by developers.  
+
+Internally to the toolkit shaders are called Programs, but that's far too confusing at this level.  Are we referring to a GPU program or a program the user writes?  Best to stick with the term shader.
+
+The shaders need a representation visible outside the toolkit and we need to change the way they’re handled inside the toolkit.  We need to be able to replace them, add brand new ones, and allow developers to control their shaders.
+
+#### The way it is now
+
+Individual shaders are represented by an object class with massive amounts of template logic and a minimum of in-line comments.  They’re opaque from the outside and only controllable through vector tile data and styles. 
+
+The [render_raster_layer](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/layers/render_raster_layer.cpp), as an example, asks for the instantiation of (eventually) RasterProgram.  Rather than having the source for the program, that then pokes into a compressed chunk of memory that contains the source, which is then uncompressed and fed into OpenGL for compilation.
+
+All of this makes sense in context, but is completely inaccessible outside the toolkit and rather incomprehensible when making changes.
+
+#### Required changes
+To let developers control their own shaders we’ll need to:
+- Allow shaders to be instantiated with shader source (one option).
+- Allow shaders to be instantiated by reference to unique names in pre-compiled file (Metal does this).
+- Set uniform values by name
+- Set uniform structs by name (Metal uses structs)
+- Tear down the specific shader (Program) definitions and replace them with one generic one.  This is almost the case now, with much of the key functionality being near the top of the class hierarchy.
+- Support for pure calculation shaders.  Useful for a variety of things, but think particles.
```

Comment:
Calculation shaders are a simple feature to add while we're working on it.  Difficult to add later.  So yes, we could skip it, but now's the cheap time to do it.

> I think we should treat heatmap as a black box for now that each backend can implement, and then we can specify a more reusable "compute" abstraction later (as it's very hard to design this without excluding many many users).

I disagree.  Heatmaps can be done with a properly formatted target.  This can be made generic within the context of OpenGL ES 2.x if need be.  It's not much of a stretch.

Reply:
Going to close this one.  Let's have the more fundamental conversation below.